dfdaemon/proxy/proxy.go

```diff
@@ -0,0 +1,117 @@
+package proxy
+
+import (
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"time"
+
+	"github.com/dragonflyoss/Dragonfly/dfdaemon/config"
+	"github.com/dragonflyoss/Dragonfly/dfdaemon/handler"
+	"github.com/sirupsen/logrus"
+)
+
+// New returns a new transparent proxy with the given rules
+func New(rules []*config.Proxy) (*TransparentProxy, error) {
+	proxy := &TransparentProxy{}
+	if err := proxy.SetRules(rules); err != nil {
+		return nil, fmt.Errorf("invalid rules: %v", err)
+	}
+	return proxy, nil
+}
+
+// TransparentProxy is an http proxy handler. It proxies requests with dfget
+// if any defined proxy rules is matched
+type TransparentProxy struct {
+	rules []*config.Proxy
+}
+
+// SetRules change the rule lists of the proxy to the given rules
+func (proxy *TransparentProxy) SetRules(rules []*config.Proxy) error {
+	proxy.rules = rules
+	return nil
+}
+
+// ServeHTTP implements http.Handler.ServeHTTP
+func (proxy *TransparentProxy) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if r.Method == http.MethodConnect {
+		HandleHTTPS(w, r)
+	} else {
+		proxy.handleHTTP(w, r)
+	}
+}
+
+func (proxy *TransparentProxy) handleHTTP(w http.ResponseWriter, req *http.Request) {
+	rt := handler.NewDFRoundTripper(nil)
+	rt.ShouldUseDfget = proxy.ShouldUseDfget
+	// delete the Accept-Encoding header to avoid returning the same cached
+	// result for different requests
+	req.Header.Del("Accept-Encoding")
+	resp, err := rt.RoundTrip(req)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusServiceUnavailable)
+		return
+	}
+	defer resp.Body.Close()
+	copyHeader(w.Header(), resp.Header)
+	w.WriteHeader(resp.StatusCode)
+	io.Copy(w, resp.Body)
+}
+
+// ShouldUseDfget returns whether we should use dfget to proxy a request. It
+// also change the scheme of the given request if the matched rule has
+// UseHTTPS = true
+func (proxy *TransparentProxy) ShouldUseDfget(req *http.Request) bool {
+	if req.Method != http.MethodGet {
+		return false
+	}
+
+	for _, rule := range proxy.rules {
+		if rule.Match(req.URL.String()) {
+			if rule.UseHTTPS {
+				req.URL.Scheme = "https"
+			}
+			return !rule.Direct
+		}
+	}
+	return false
+}
+
+// HandleHTTPS handles a CONNECT request and proxy an https request through an
+// http tunnel.
+func HandleHTTPS(w http.ResponseWriter, r *http.Request) {
```

Comment:
The same as above, why use capitalization here?

Reply:
I was thinking about using this in dfdaemon/handler/root_handler.go to fix the old reverse proxy, so I made this exported. But now I think maybe it's better to leave the old reverse proxy as it is, so the user will get errors if they used the wrong ports for proxy. I'll make this un-exported.

Reply:
Fixed in the latest commit